src/umpire/util/AllocationRecordMap.cpp

```diff
@@ -0,0 +1,387 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2018-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory
+//
+// Created by David Beckingsale, david@llnl.gov
+// LLNL-CODE-747640
+//
+// All rights reserved.
+//
+// This file is part of Umpire.
+//
+// For details, see https://github.com/LLNL/Umpire
+// Please also see the LICENSE file for MIT license.
+//////////////////////////////////////////////////////////////////////////////
+#include "umpire/util/AllocationRecordMap.hpp"
+
+#include "umpire/util/FixedMallocPool.hpp"
+
+#include "umpire/util/Macros.hpp"
+
+#include <iterator>
+#include <sstream>
+
+namespace umpire {
+namespace util {
+
+namespace {
+
+template <typename T>
+struct ListBlock
+{
+  T rec;
+  ListBlock* prev;
+};
+
+static umpire::util::FixedMallocPool block_pool(sizeof(ListBlock<AllocationRecord>));
+
+// Forward declare const iterator
+class RecordListConstIterator;
+
+// TODO Profile this to determine if we should add a small string
+// optimization on top of the block pooling.
+class RecordList
```

Comment:
Where is thread safety insured for this object?

Reply:
After reading further down into the public section, it appears that mutual exclusion is being set up there.  Let me know if this is not right, otherwise, you should be able to resolve this comment.

Reply:
Indeed. This is only used inside `AllocationRecordMap`, which uses a lock.